Hypcond

n2owens opened the pull request:
Hyperbolic conduction code, imported from gasoline

Commit:
Hyperbolic conduction code ported from Gasoline. To run, configure with --enable-hypcond
ACCZERO ported from Gasoline.
ThermalDiffOnly added for running tests on conduction without spitzer.

n2owens committed:
Delete charmrun.smp

n2owens committed:
Delete settings

trquinn commented on ParallelGravity.cpp:
```diff
@@ -2508,6 +2508,10 @@ void Main::setupICs() {
 
           treeProxy.loadTipsy(basefilename, dTuFac, bInDPos, bInDVel,
               CkCallbackResumeThread());
+#ifdef HYPCOND
```
Having this initialization here instead of after the try block means that initQCond() won't happen for NChilada ICs.  Is this intentional?

trquinn commented on Sph.cpp:
```diff
@@ -146,6 +146,25 @@ DataManager::initCooling(double dGmPerCcUnit, double dComovingGmPerCcUnit,
     contribute(cb);
     }
 
+//read QCond at beginning of run
```
See the coolontime reading code a few lines above to see how to generalize this for NChilada files.

n2owens replied:
I have updated this and the part regarding the previous comment. I am unfamiliar with NChilada files, though, so I have not been able to test this.

trquinn commented on Sph.cpp:
```diff
@@ -1384,13 +1431,24 @@ void TreePiece::getAdiabaticGasPressure(double gamma, double gammam1, double dTu
         double fThermalCond = dThermalCondCoeff*pow(p->uPred(),2.5);
         double fThermalCond2 = dThermalCond2Coeff*pow(p->uPred(),0.5);
+#ifdef HYPCOND
+        p->dThermalCondTau() = fThermalCond/p->fDensity/(p->c()*p->c() + 0.0000001); /*NAO HYP*/
```
I'm guessing that the 1e-7 is to avoid divide by zeros.  It's a little dodgy since p->c() is a dimensional quantity, so the 1e-7 could be significant depending on the units.

n2owens replied:
I have changed this to 1e-33, which is probably a better choice. 
This has also been moved into the Sph Pressure terms part of the code, so that thermalcondtau can be in the param file.

trquinn replied:
See more recent review.

trquinn commented on Sph.cpp:
```diff
@@ -1811,6 +2048,11 @@ void updateParticle(GravityParticle *a, GravityParticle *b,
                         * massDiffFac(b);
+                #ifdef HYPCOND
+                a->dQCondDot() +=  params->diffq * bParams->rNorm * massDiffFac(b);
+                if (a->iOrder == 34374) fprintf(stderr,"iOrder: %i\tQDot: %g\n\n", a->iOrder, a->dQCondDot()[0]);
```
Debugging statement; should be removed.

n2owens committed:
Removed debug settings including removing print statement and adding
missing tinv variable back into Sph.cpp

Modify Reading QCond files so that code should work properly with
NChilada files

Updated to more modern version of the the code where thermalcondtau is
now declared in the .param file

Update timestep criteria and artificial diffusion to better match modern
version of GASOLINE code

trquinn committed:
Merge branch 'master' into hypcond

trquinn committed:
Struct parameters: remove HYPCOND ifdef and add PUP.

This allows this branch to compile when HYPCOND is not enabled.

trquinn commented on ParallelGravity.cpp:
```diff
@@ -3732,6 +3748,9 @@ void Main::writeOutput(int iStep)
            * (multiphase properties and effective temperature)
            */
+#ifdef HYPCOND
+              HypCondOutputParams pdQCondOut(achFile,param.iBinaryOut,0.0);
```
The last argument to the constructor is the output time (dOutTime).
Note that all the SUPERBUBBLE output constructors also have this mistake.

trquinn committed:
Main::writeOutput(): use dOutTime for superbubble and hypcond

trquinn committed:
HypGrads::fcnSmooth(): get rid of unused variables

Also clean up iOrder formatting.

trquinn commented on configure.ac:
```diff
@@ -298,7 +298,23 @@ fi
+ARG_ENABLE([acczero], [zero acceleration], [FLAG_ACCZERO], [-DACCZERO], [no])
+
+
+#THERMALDIFFONLY
```
What is this for?  A slightly longer explanation would help.

trquinn commented on Sph.cpp:
```diff
@@ -1418,13 +1479,18 @@ void TreePiece::getAdiabaticGasPressure(double gamma, double gammam1, double dTu
         p->PoverRho2() = PoverRho/p->fDensity;
         double Tp = p->uPred() / dTuFac;
+#ifdef THERMALDIFFONLY
+        double fThermalCond = 0.5;
```
What is the meaning of these values?

trquinn commented on Sph.cpp:
```diff
@@ -1829,6 +2048,43 @@ void PressureSmoothParams::fcnSmooth(GravityParticle *p, int nSmooth,
  * @param sign 1 for a = p (the self particle) and -1 for a = q (the neighbor)
  */
+
+void getInitialGrad(GravityParticle *a, GravityParticle *b,
```
The use of PressSmoothParticle is overkill: only rNorm is used.
Also: sign is unused.

trquinn replied:
This code is now gone.

trquinn commented on Sph.cpp:
```diff
@@ -1532,6 +1598,91 @@ void TreePiece::getCoolingGasPressure(double gamma, double gammam1, double dTher
+        params.gradQ1x = -(p->dQCondPred()[0] - q->dQCondPred()[0])*params.dx*dIHarmDens;
+        params.gradQ1y = -(p->dQCondPred()[1] - q->dQCondPred()[1])*params.dx*dIHarmDens;
+        params.gradQ1z = -(p->dQCondPred()[2] - q->dQCondPred()[2])*params.dx*dIHarmDens;
+        }
```
This brace is missplaced!  Only one neighbor particle is considered!

trquinn replied:
more code that is now gone.

trquinn commented on Sph.cpp:
```diff
@@ -1760,27 +1926,80 @@ void PressureSmoothParams::fcnSmooth(GravityParticle *p, int nSmooth,
+                            double fDensAvg = 0.5*(p->fDensity+q->fDensity) ;
+                            double havrg = 0.5*(ph+0.5*q->fBall);
+                            double csavrg0 = 0.5*(p->c()+q->c());
+                            double csavrg1 = 1.0*fThermalCondAvrg/fDensAvg/havrg ;
```
double divides should be replaced by division by a product for both performance and readability.

trquinn replied:
The 1.0 here is superfluous.

trquinn replied:
I'm having trouble working out how this can have units of velocity.  Perhaps a reference to equations in your paper would help.

trquinn commented on Sph.cpp:
```diff
@@ -1760,27 +1926,80 @@ void PressureSmoothParams::fcnSmooth(GravityParticle *p, int nSmooth,
+                            double csavrg1 = 1.0*fThermalCondAvrg/fDensAvg/havrg ;
+                            double csavrg = min(csavrg0, csavrg1);
+                            dHypStable = havrg*csavrg/fDensAvg;
+                            diffuHyp = 1.0/p->fDensity/q->fDensity;
```
double divides should be replaced by division by a product for both performance and readability.

trquinn commented on Sph.cpp:
```diff
@@ -1760,27 +1926,80 @@ void PressureSmoothParams::fcnSmooth(GravityParticle *p, int nSmooth,
+                               }
+                            #else
+                            if (dThermalCond > 0 && (dt_diff = 0.4*ph
+                                   /(sqrt(p->fThermalCond()/p->fDensity/p->dThermalCondTau()) )) < dt) {
```
double divides should be replaced by division by a product for both performance and readability.

trquinn commented on Sph.cpp:
```diff
@@ -1760,27 +1926,80 @@ void PressureSmoothParams::fcnSmooth(GravityParticle *p, int nSmooth,
+                        dHypCond = 0.0;
+                        #endif
                     #endif
+                    #ifdef HYPCOND
```
Can this line be outside the ifdef to save duplicating it on line 2000?

trquinn commented on Sph.h:
```diff
@@ -206,12 +208,83 @@ class PressureSmoothParams : public SmoothParams
+class HypGrads : public SmoothParams
+{
+    double dTime;
+    double a, H; // Cosmological parameters
```
There are a lot of unused attributes in this class.  Are the cosmology parameters needed?

trquinn commented on Sph.cpp:
```diff
@@ -1829,6 +2048,43 @@ void PressureSmoothParams::fcnSmooth(GravityParticle *p, int nSmooth,
+#endif
+}
+
+void minMod(double val1, double val2, double &slope_1,double &slope_2) {
```
This function needs documentation.  At least: what do all the parameters mean?

trquinn commented on Sph.cpp:
```diff
@@ -1652,7 +1812,13 @@ void PressureSmoothParams::fcnSmooth(GravityParticle *p, int nSmooth,
-
+#ifdef HYPCOND
+#ifndef THERMALDIFFONLY
+    p->dThermalCondTau() = dThermalCondTau*p->fThermalCond()/p->fDensity/(p->c()*p->c() + 1e-33); /*NAO HYP*/
```
We now have a floor on "u" so c() should never be zero.  I suggest removing the 1e-33 and putting a CkAssert(p->c() > 0.0) before this line.

trquinn commented on Sph.cpp:
```diff
@@ -1760,27 +1926,80 @@ void PressureSmoothParams::fcnSmooth(GravityParticle *p, int nSmooth,
+                    double gradQzp = dot(p->gradQz(),params.dx); double gradQzq = dot(q->gradQz(),params.dx);
+                    minMod(p->dQCond()[2], q->dQCondPred()[2], gradQzp, gradQzq);
+                    params.diffu = diffTh*(p->uPred()-q->uPred());
+                    if ((fabs(gradUp)<1e-20)||(fabs(gradUq)<1e-20))
```
Where does 1e-20 come from?  Why is this being done?

trquinn commented on Sph.cpp:
```diff
@@ -1760,27 +1926,80 @@ void PressureSmoothParams::fcnSmooth(GravityParticle *p, int nSmooth,
+                    if ((fabs(gradUp)<1e-20)||(fabs(gradUq)<1e-20))
+                        params.diffu += 0.5*dHypStable*(p->uPred() - q->uPred() + 0.5*gradUp + 0.5*gradUq);
+                    params.diffuHypCond = 0.0;
+                    params.diffuHypCond += diffuHyp*(p->dQCondPred()[0] - q->dQCondPred()[0] + 0.5*gradQxp + 0.5*gradQxq)*params.dx[0];
```
This looks like another dot product.  Why not write it that way: turn gradQ[xyz]p and gradQ[xyz]q into Vector3D<double>

trquinn reviewed (changes requested):
I gave it a more careful read, and have a few suggested changes, including fixing an obvious bug.

Is there a test simulation we can use to check this?

n2owens committed:
Gradient projection script removed (outdated).

When compiling with THERMALDIFFONLY, dThermalCondCoeff now specifies the
constant conduciton coefficient.

Hyperbolic conduction now removes saturation (not needed).

Comments added for clarity.

Minor bug fixes.

n2owens commented on TreePiece.cpp:
```diff
@@ -1601,11 +1601,12 @@ void TreePiece::kick(int iKickRung, double dDelta[MAXRUNG+1],
+                  Vector3D<double> QdotCorrected;
+                  QdotCorrected = (p->dQCondDot() - p->dQCond())*tinv;
+                  p->dQCond() +=  QdotCorrected*duDelta[p->rung];
+                  p->dQCondPred() = p->dQCond();
```
In GASOLINE, we produce the corrected Qdot in a separate function, post SPH pressure terms. 
I was wondering if we should have something similar, or if doing the semi-implicit step here works.
